File: cloudstairs/server/src/main/java/com/taot/cloudstairs/util/AesUtil.java
package com.taot.cloudstairs.util;

import java.security.SecureRandom;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;

import org.apache.commons.codec.binary.Hex;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AesUtil {
    
    private static Logger logger = LoggerFactory.getLogger(AesUtil.class);

    private static SecureRandom rand = new SecureRandom();
    
    public static final int BLOCK_SIZE = 16;

    private AesUtil() {
    }
    
    public static byte[] randomAesKey() {
        byte[] bytes = new byte[BLOCK_SIZE];
        synchronized (rand) {
            rand.nextBytes(bytes);
        }
        return bytes;
    }
    
    public static byte[] decrypt(SecretKey key, byte[] encryptedData) {
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/NoPadding");
            logger.info("AES Key: " + Hex.encodeHexString(key.getEncoded()));
            
            IvParameterSpec iv = new IvParameterSpec(encryptedData, 0, BLOCK_SIZE);
            
            logger.info("encryptedData.length: " + encryptedData.length);
            cipher.init(Cipher.DECRYPT_MODE, key, iv);
            byte[] data = cipher.doFinal(encryptedData, BLOCK_SIZE, encryptedData.length - BLOCK_SIZE);
            return data;
        } catch (Exception e) {
            throw new RuntimeException("Error decrypting AES. " + e.getMessage(), e);
        }
    }
}

File: cloudstairs/server-play/app/util/AesUtil.scala
package util

import java.security.SecureRandom
import javax.crypto.{ Cipher, SecretKey }
import javax.crypto.spec.IvParameterSpec
import org.apache.commons.codec.binary.Hex
import org.slf4j.{ Logger, LoggerFactory }


object AesUtil extends Logging {

  val BLOCK_SIZE: Int = 16

  private val rand = new SecureRandom

  def randomAesKey(): Array[Byte] = {
    val bytes = new Array[Byte](BLOCK_SIZE)
    synchronized {
      rand.nextBytes(bytes)
    }
    logger.debug("Random aes key (hex): " + Hex.encodeHexString(bytes))
    bytes
  }

  def decrypt(key: SecretKey, encrypted: Array[Byte]): Array[Byte] = {
    val cipher = Cipher.getInstance("AES/CBC/NoPadding")
    logger.info("AES Key: " + Hex.encodeHexString(key.getEncoded))
    val iv = new IvParameterSpec(encrypted, 0, BLOCK_SIZE)
    logger.debug("encoded.length: " + encrypted.length)
    cipher.init(Cipher.DECRYPT_MODE, key, iv)
    val data = cipher.doFinal(encrypted, BLOCK_SIZE, encrypted.length - BLOCK_SIZE)
    data
  }
}
